server/channels/jobs/batch_worker.go

```diff
@@ -73,10 +84,14 @@ func (worker *BatchWorker) Run() {
 
 // Stop interrupts the worker even if the migration has not yet completed.
 func (worker *BatchWorker) Stop() {
```

Comment:
Longer term effort, but I'd love to shift to a model where a worker can only be `Run` once, after which you need a new worker. Presumably, this would only happen on startup or if the job subsystem was "rebooted" due to a major configuration change. I sense this would substantially simplify the logic here?

Reply:
Hmm .. I think reusing the same worker is a fairly well-established pattern. What we really need to do is make `Run` non-blocking and move the main for-select loop to a separate goroutine from inside `Run`. That simplifies the caller side and would allow to refactor some code. Other than that, it's not that complicated IMO.

Reply:
> I think reusing the same worker is a fairly well-established pattern.

I think this is different than only calling `Run` once. The expected lifecycle of a worker is to call `Run` once and have it process from the queue of jobs until the server shuts down. In this sense, I agree that reusing the same worker is a well-established pattern that we should keep.

But in addition to the queue, we also currently support restarting a worker in place. This isn't necessary for the logic of the queue: I think it's only there as an artifact of the way the job subsystem has been designed. When we do need to stop -- exceptional events, like shutting down or reconfiguring the job subsystem -- it seems viable to just wind down the workers and clean up completely. Starting from that point would just be equivalent to what happens when the server starts.

Thoughts @hanzei & @agnivade?

Reply:
I see your point. Yeah I think that would work as well, and might simplify some of the stuff here.

Reply:
In that case, we might also need to handle the edge-case of one worker shutting down and another worker starting. Although there won't be a data-race, but there might be a logic race of two separate jobs being in a weird state. We'll need to look into that as well.